torch/_inductor/graph.py

```diff
@@ -181,8 +216,159 @@ def __init__(
+            log.debug(
+                "SKIP LAYOUT OPT BECAUSE SOME CONVOLUTTION HAS SMALLER OUT_CHANNEL"
+            )
+            config.layout_opt = False
+            return True
+
+        # Following models are skipped due to this:
+        # - functorch_maml_omniglot
+        if all(
+            n.args[1].meta["val"].size(0) <= 64 and n.args[1].meta["val"].size(1) <= 64
+            for n in conv_nodes
+        ):
+            log.debug("SKIP LAYOUT OPT BECAUSE ALL CONVOLUTION CHANNELS TOO SMALL")
+            config.layout_opt = False
+            return True
+
+        # aten._scaled_dot_product_flash_attention requires the last stride of query/key/value
+        # to be 1. Check https://gist.github.com/shunting314/fa6eeab2aad8d1265c4d5e50b560d94f
+        # for more details.
+        #
+        # When a model contains aten._scaled_dot_product_flash_attention and we enable layout optimization,
+        # the op may get channels last input and fail. Example include: twins_pcpvt_base, xcit_large_24_p8_224
+        #
+        # We disable layout optimization if a model contains aten._scaled_dot_product_flash_attention.
+        #
+        # An alternative is to do necessary layout convertion to make sure aten._scaled_dot_product_flash_attention's
+        # inputs have the layout needed. But that seems to have worse perf than disabing the layout opt.
+        for n in gm.graph.nodes:
+            if n.target == torch.ops.aten._scaled_dot_product_flash_attention.default:
+                log.debug(
+                    "SKIP LAYOUT OPT BECAUSE _scaled_dot_product_flash_attention found"
+                )
+                config.layout_opt = False
+                return True
+
+        return False
+
+    def find_nodes_prefer_channels_last(self):
+        """
+        The rule to decide if an node prefer channels last is simple.
+        1. if it's input/output of a convolution
+        2. if one of its user preferes channels last
+
+        We have rule 1 because cudnn runs a faster convolution kernel for channels last inputs;
+        Rule 2 is also important. It makes sure that indirect inputs to convolution also prefers
+        channels last.
+
+        Consider the scenario: conv -> batch-norm -> relu -> conv
+        Without rule 2, batch-norm output may use a congituous layout. That will cause 2 extra copies:
+        1. the output of batch-norm should be channels last initially since its input is a conv's output.
+           Forcing the batch-norm's output to be contiguous results in the first copy
+        2. The second conv's input is initially contiguous. This layout is propagated from the batch-norm's output.
+           We need convert it to channels last layout which results in the second copy.
+        With rule 2, we makes sure all the tensors in the chain uses channels last layout. So both copies
+        can be saved.
+        """
+        output_set = set()
+        for n in reversed(self.module.graph.nodes):
```

Comment:
Because we want to collect convolution and its transitive input nodes and use channels last layout for them.

If we go in normal order, we would need do a search thru the graph to decide if a node can reach a convolution;
go in reverse order simplify this a lot since we can start with convolution and propagate thru its inputs as we iterate thru the reversed node list.